File: moe/tools/moe.generator.project/src/main/resources/org/moe/generator/project/templates/master-detail/java/ui/DetailViewController.java
package ${_PLACEHOLDER_PACKAGE_NAME}.ui;

import apple.uikit.UILabel;
import apple.uikit.UIViewController;
import org.moe.natj.general.Pointer;
import org.moe.natj.general.ann.Owned;
import org.moe.natj.general.ann.RegisterOnStartup;
import org.moe.natj.objc.ObjCRuntime;
import org.moe.natj.objc.ann.IBOutlet;
import org.moe.natj.objc.ann.ObjCClassName;
import org.moe.natj.objc.ann.Property;
import org.moe.natj.objc.ann.Selector;

@org.moe.natj.general.ann.Runtime(ObjCRuntime.class)
@ObjCClassName("DetailViewController")
@RegisterOnStartup
public class DetailViewController extends UIViewController {

    @Owned
    @Selector("alloc")
    public static native DetailViewController alloc();

    @Selector("init")
    public native DetailViewController init();

    protected DetailViewController(Pointer peer) {
        super(peer);
    }

    private String detailItem;

    public String getDetailItem() {
        return detailItem;
    }

    public void setDetailItem(String detailItem) {
        this.detailItem = detailItem;
        configureView();
    }

    @Override
    public void viewDidLoad() {
        super.viewDidLoad();
        configureView();
    }

    private void configureView() {
        if (detailItem != null && detailDescriptionLabel() != null) {
            detailDescriptionLabel().setText(detailItem);
        }
    }

    @Property
    @IBOutlet
    @Selector("detailDescriptionLabel")
    public native UILabel detailDescriptionLabel();

    @Selector("setDetailDescriptionLabel:")
    public native void setDetailDescriptionLabel(UILabel detailDescriptionLabel);
}

File: tools/moe.generator.project/src/main/resources/org/moe/generator/project/templates/master-detail/kotlin/ui/DetailViewController.kt
package ${_PLACEHOLDER_PACKAGE_NAME}.ui

import apple.uikit.UILabel
import apple.uikit.UIViewController
import org.moe.natj.general.Pointer
import org.moe.natj.general.ann.Owned
import org.moe.natj.general.ann.RegisterOnStartup
import org.moe.natj.objc.ObjCRuntime
import org.moe.natj.objc.ann.IBOutlet
import org.moe.natj.objc.ann.ObjCClassName
import org.moe.natj.objc.ann.Property
import org.moe.natj.objc.ann.Selector

@org.moe.natj.general.ann.Runtime(ObjCRuntime::class)
@ObjCClassName("DetailViewController")
@RegisterOnStartup
class DetailViewController protected constructor(peer: Pointer) : UIViewController(peer) {

    @Selector("init")
    override external fun init(): DetailViewController

    var detailItem: String? = null
        set(detailItem) {
            field = detailItem
            configureView()
        }

    override fun viewDidLoad() {
        super.viewDidLoad()
        configureView()
    }

    private fun configureView() {
        if (this.detailItem != null && detailDescriptionLabel() != null) {
            detailDescriptionLabel()!!.setText(this.detailItem)
        }
    }

    @Property
    @IBOutlet
    @Selector("detailDescriptionLabel")
    external fun detailDescriptionLabel(): UILabel?

    @Selector("setDetailDescriptionLabel:")
    external fun setDetailDescriptionLabel(detailDescriptionLabel: UILabel?)

    companion object {

        @Owned
        @Selector("alloc")
        @JvmStatic external fun alloc(): DetailViewController
    }
}
